### PR TITLE
feat: flox update, introduce global lock

### PR DESCRIPTION
**Pull request by mkenigs**
- Introduce a global manifest lockfile.
- Add an update subcommand that either:
  - Updates an environments inputs.
  - Updates inputs in the global manifest, storing the lock in the newly introduced global lockfile.

Code notes:
- The `pkgdb_errors` module was renamed to `pkgdb`, and a `call_pkgdb` function was factored out along with some errors.
- The manifest argument to all pkgdb subcommands was made optional to allow them to operate on the global manifest. Commands will still fail unless an empty manifest is explicitly set in the implementation of the command, which is the approach taken by search, and by update in this commit.
- `transact_with_lockfile_contents` was added to `CoreEnvironment`

### CODE REVIEW

File: pkgdb/src/resolver/mixins.cc
Comment (mkenigs) [inline pkgdb/src/resolver/mixins.cc:291]: I wasn't sure if this was the best way to make this change. I think most of the `Environment` class is not written with an optional manifest in mind. But it seems to work, and it seems like `search` uses an empty manifest just fine. Not sure if there's a better approach now, or maybe we can just clean up in the future

File: pkgdb/src/resolver/mixins.cc
Comment (mkenigs) [inline pkgdb/src/resolver/mixins.cc:306]: I'm thinking there can only be one lockfile. And `flox` should decide which to use rather than `pkgdb`. Wanted to flag as I'm not sure we've really discussed this behavior. And search and show will have to be updated to use this lockfile.

File: cli/flox-rust-sdk/src/models/environment/core_environment.rs
Comment (aakropotkin) [inline cli/flox-rust-sdk/src/models/environment/core_environment.rs:351]:
```suggestion
    /// so that calling `pkgdb manifest lock` with the new lockfile_contents is
```

File: pkgdb/src/resolver/mixins.cc
Comment (aakropotkin) [reply pkgdb/src/resolver/mixins.cc:291]: Yeah we should probably make that member optional in another PR soon. It's been kind of wonky having to pass an empty manifest manually.

This works for now though.

File: pkgdb/src/resolver/mixins.cc
Comment (aakropotkin) [reply pkgdb/src/resolver/mixins.cc:306]: For cases where the global manifest has inputs that don't exist in the project we'll want two; but for now that approach for a single lock sounds good to me 👍

**Review by aakropotkin — CHANGES_REQUESTED** [review]
LGTM other than that one typo fix.

Ping me for a quick approval after it's applied ✌️

File: pkgdb/src/resolver/mixins.cc
Comment (mkenigs) [reply pkgdb/src/resolver/mixins.cc:306]: I thought in that case we were copying the input into the environment, so that if it gets shared, whoever pulls the environment can still build and update it?

File: cli/flox-rust-sdk/src/models/pkgdb.rs
Comment (mkenigs) [inline cli/flox-rust-sdk/src/models/pkgdb.rs:1]: Not sure if this should be in models

File: cli/flox-rust-sdk/src/models/environment/mod.rs
Comment (mkenigs) [reply cli/flox-rust-sdk/src/models/environment/mod.rs:102]: As the command is `flox update` I don't think `update_inputs` is much clearer. Fixing the async...

Comment (ysndr) [reply cli/flox-rust-sdk/src/models/environment/mod.rs]: Hmm might be just me finding "update" a bit ambiguous
Don't block on it if you prefer to leave it

Comment (mkenigs) [reply cli/flox-rust-sdk/src/models/environment/mod.rs]: 👍 won't block on it then

Comment (ysndr) [reply cli/flox-rust-sdk/src/models/environment/mod.rs]: Just remove the async :)

File: cli/flox-rust-sdk/src/models/pkgdb.rs
Comment (mkenigs) [reply cli/flox-rust-sdk/src/models/pkgdb.rs:9]: Done

File: cli/flox-rust-sdk/src/models/environment/mod.rs
Comment (mkenigs) [reply cli/flox-rust-sdk/src/models/environment/mod.rs]: Done

File: cli/flox/src/commands/environment.rs
Comment (mkenigs) [reply cli/flox/src/commands/environment.rs]: Yep might be cleaner, didn't do it yet cause I thought it would take me longer.

**Review by aakropotkin — APPROVED** [review]

**Review by ysndr — APPROVED** [review]

**Pull request merged**
